src/fam.rs

```diff
@@ -0,0 +1,695 @@
+///
+/// This is unsafe due to the number of constraints that aren't checked:
+/// * the implementer should be a POD
+/// * the implementor should contain a flexible array member of elements of type `Entry`
+/// * `Entry` should be a POD
+///
+/// Violating these may cause problems.
+///
+/// # Example
+///
+/// ```
+/// use vmm_sys_util::fam::*;
+///
+/// #[repr(C)]
+/// #[derive(Default)]
+/// pub struct __IncompleteArrayField<T>(::std::marker::PhantomData<T>, [T; 0]);
+/// impl<T> __IncompleteArrayField<T> {
+///     #[inline]
+///     pub fn new() -> Self {
+///         __IncompleteArrayField(::std::marker::PhantomData, [])
+///     }
+///     #[inline]
+///     pub unsafe fn as_ptr(&self) -> *const T {
+///         ::std::mem::transmute(self)
+///     }
+///     #[inline]
+///     pub unsafe fn as_mut_ptr(&mut self) -> *mut T {
+///         ::std::mem::transmute(self)
+///     }
+///     #[inline]
+///     pub unsafe fn as_slice(&self, len: usize) -> &[T] {
+///         ::std::slice::from_raw_parts(self.as_ptr(), len)
+///     }
+///     #[inline]
+///     pub unsafe fn as_mut_slice(&mut self, len: usize) -> &mut [T] {
+///         ::std::slice::from_raw_parts_mut(self.as_mut_ptr(), len)
+///     }
+/// }
+///
+/// impl<T> ::std::clone::Clone for __IncompleteArrayField<T> {
+///     #[inline]
+///     fn clone(&self) -> Self {
+///         Self::new()
+///     }
+/// }
+///
+/// #[repr(C)]
+/// #[derive(Clone, Default)]
+/// struct MockFamStruct {
+///     pub len: u32,
+///     pub padding: u32,
+///     pub entries: __IncompleteArrayField<u32>,
+/// }
+///
+/// unsafe impl FamStruct for MockFamStruct {
+///     type Entry = u32;
+///
+///     fn len(&self) -> usize {
+///         self.len as usize
+///     }
+///
+///     fn set_len(&mut self, len: usize) {
+///         self.len = len as u32
+///     }
+///
+///     fn max_len() -> usize {
+///         100
+///     }
+///
+///     fn as_slice(&self) -> &[u32] {
+///         let len = self.len();
+///         unsafe { self.entries.as_slice(len) }
+///     }
+///
+///     fn as_mut_slice(&mut self) -> &mut [u32] {
+///         let len = self.len();
+///         unsafe { self.entries.as_mut_slice(len) }
+///     }
+/// }
+///
+/// type MockFamStructWrapper = FamStructWrapper<MockFamStruct>;
+/// ```
+///
+#[allow(clippy::len_without_is_empty)]
+pub unsafe trait FamStruct {
+    /// The type of the zero length array entries
+    type Entry: PartialEq + Copy;
+
+    /// Get the array length
+    fn len(&self) -> usize;
+
+    /// Set the array length
+    fn set_len(&mut self, len: usize);
+
+    /// Get max array length
+    fn max_len() -> usize;
+
+    /// Get the array entries as slice
+    fn as_slice(&self) -> &[Self::Entry];
+
+    /// Get the array entries as mut slice
+    fn as_mut_slice(&mut self) -> &mut [Self::Entry];
+}
+
+/// A wrapper that helps in treating a [`FamStruct`](trait.FamStruct.html) similarly to
+/// an actual `Vec`.
+///
```

Comment:
Remove this line if no other description.

Reply:
Also, could you please kindly separate the documentation into a short summary line of its functionality on the first line, and more explanation in a new paragraph? Ditto for others below.

Reply:
Done